wit/network.wit

```diff
@@ -13,11 +13,77 @@ default interface network {
 	drop-network: func(this: network)
 
 
-
+	/// Error codes.
+	/// 
+	/// In theory, every API can return any error code.
+	/// In practice, API's typically only return the errors documented per API
+	/// combined with a couple of errors that are always possible:
+	/// - `unknown`
+	/// - `access-denied`
+	/// - `not-supported`
+	/// - `resource-limit-reached`
+	/// - `out-of-memory`
+	/// 
+	/// See each individual API for what the POSIX equivalents are. They sometimes differ per API.
 	enum error {
+		// ### GENERAL ERRORS ###
+
+		/// Unknown error
 		unknown,
-		again,
-		// TODO ...
+
+		/// POSIX equivalent: EACCES, EPERM
+		access-denied,
+
+		/// POSIX equivalent: EOPNOTSUPP
+		not-supported,
+
+		/// POSIX equivalent: EMFILE, ENFILE, ENOMEM, ENOBUFS, EAI_MEMORY
+		resource-limit-reached,
+
+		/// POSIX equivalent: ENOMEM, ENOBUFS, EAI_MEMORY
+		out-of-memory,
+
+		timeout,
+		concurrency-conflict,
+
+
+		// ### IP ERRORS ###
+		address-family-not-supported,
+		address-family-mismatch,
+		invalid-remote-address,
+		ipv4-only-operation,
+		ipv6-only-operation,
+
+
+
+		// ### TCP & UDP SOCKET ERRORS ###
+		already-attached,
+		already-bound,
+		already-connected,
+		not-bound,
+		not-connected,
+		address-not-bindable,
+		address-in-use,
+		ephemeral-ports-exhausted,
+		remote-unreachable,
+		
+
+		// ### TCP SOCKET ERRORS ###
+		already-listening,
+		not-listening,
+		connection-refused,
+		connection-reset,
+		
+		
+		// ### UDP SOCKET ERRORS ###
+		datagram-too-large,
+
+
+		// ### NAME LOOKUP ERRORS ###
+		invalid-name,
+		name-unresolvable,
+		temporary-resolver-failure,
+		permanent-resolver-failure,
```

Comment:
Would it make sense to move resolver errors to a separate enum? At a glance, the only code they share with the socket errors is `out-of-memory`.

Reply:
They also share:
- `address-family-not-supported`
- `access-denied`
- `not-supported`
- `resource-limit-reached`

Reply:
When would a resolver API return `access-denied`? When would it return `not-supported`?

Reply:
Examples:
- `access-denied`: The network does not allow you to look up the specific `name`.
- `not-supported`: The network does not support name resolution.

I don't know operation systems ever return these error codes natively.
But in WASI, with its increased focus on sandbox-ability and virtualization, it makes sense to me to give embedders an escape hatch in case they don't want to support/give access to a specific function in an interface. This applies in general, not just the resolve API.